### ISSUE TITLE
Is batchnorm working?

### ISSUE THREAD

**Issue by memoakten** [zhangqianhui/vae-gan-tensorflow#5]
Hey, thanks for this code.

I'm curious is `batch_norm` actually working? As far as I understand, you need to pass in `is_training` so it can act accordingly:

`is_training: Whether or not the layer is in training mode. In training mode it would accumulate the statistics of the moments into moving_mean and moving_variance using an exponential moving average with the given decay. When it is not in training mode then it would use the values of the moving_mean and the moving_variance.`

And also the ops it creates aren't added as dependencies to the training ops, so the moving_mean and moving_variance aren't updated in time. You'd need to add these as a dependency to your train ops manually.

```
Note: when training, the moving_mean and moving_variance need to be updated. By default the update ops are placed in tf.GraphKeys.UPDATE_OPS, so they need to be added as a dependency to the train_op. For example:
  update_ops = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
  with tf.control_dependencies(update_ops):
    train_op = optimizer.minimize(loss)
```

At least this is my understanding. Is there a reason you omitted these steps?
https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm

**Comment by zhangqianhui** [comment]
One can set updates_collections=None to force the updates in place, but that can have a speed penalty, especially in distributed settings.

**Comment by zhangqianhui** [comment]
https://www.tensorflow.org/api_docs/python/tf/contrib/layers/batch_norm

**Comment by memoakten** [comment]
Yes I saw that after posting :) (Interestingly that option has been removed from the stable API https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization )

What about the is_training flag? It defaults to true in the contrib version but is there any impact in not setting it to false during inference? The docs (from stable API,[ tf.layers.batch_normalisation](https://www.tensorflow.org/api_docs/python/tf/layers/batch_normalization)) state: `NOTE: make sure to set this parameter correctly, or else your training/inference will not work properly.`

**Comment by zhangqianhui** [comment]
https://stackoverflow.com/questions/43234667/tf-layers-batch-normalization-large-test-error

About the is_training flag, my code has some error, thanks. but, I think it don't has much impact for GAN. But for classification , we should set is_training =False during inference.
